src/dragonball/src/dbs_pci/src/msi.rs

```diff
@@ -0,0 +1,759 @@
+        let field = (self.msg_ctl >> 4) & 0x7;
+
+        if field > 5 {
+            return 0;
+        }
+
+        1 << field
+    }
+
+    /// Check whether 64-bit message address is supported or not.
+    pub fn addr_64_bits(&self) -> bool {
+        self.msg_ctl & MSI_CTL_64_BITS == MSI_CTL_64_BITS
+    }
+
+    /// Check whether per vector masking extension is supported or not.
+    pub fn per_vector_mask(&self) -> bool {
+        self.msg_ctl & MSI_CTL_PER_VECTOR == MSI_CTL_PER_VECTOR
+    }
+
+    /// Check whether the `vector` is masked or not.
+    pub fn vector_masked(&self, vector: u16) -> bool {
+        if !self.per_vector_mask() {
+            return false;
+        }
+
+        (self.mask_bits >> vector) & 0x1 == 0x1
+    }
+
+    /// Get size of the PCI MSI capability structure, including the capability header.
+    pub fn size(&self) -> u32 {
+        let mut size: u32 = 0xa;
+
+        if self.addr_64_bits() {
+            size += 0x4;
+        }
+        if self.per_vector_mask() {
+            size += 0xa;
+        }
+
+        size
+    }
+
+    /// Handle read accesses to the PCI MSI capability structure.
+    pub fn read(&mut self, offset: u64, data: &mut [u8]) -> std::io::Result<()> {
+        let (msg_data_offset, addr_hi_offset, mask_bits_offset) = self.get_offset_info();
+
+        // Please be really careful to touch code below, you should have good understanding of
+        // PCI MSI capability.
+        match data.len() {
+            1 => match offset {
+                0 => data[0] = self.cap_id_next as u8,
+                1 => data[0] = (self.cap_id_next >> 8) as u8,
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() => {
+                    data[0] = self.mask_bits as u8;
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 1 => {
+                    data[0] = (self.mask_bits >> 8) as u8;
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 2 => {
+                    data[0] = (self.mask_bits >> 16) as u8;
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 3 => {
+                    data[0] = (self.mask_bits >> 24) as u8;
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 4 => {
+                    self.get_pending_state(data, 0);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 5 => {
+                    self.get_pending_state(data, 1);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 6 => {
+                    self.get_pending_state(data, 2);
+                }
+                x if mask_bits_offset.is_some() && x == mask_bits_offset.unwrap() + 7 => {
+                    self.get_pending_state(data, 3);
+                }
```

Comment:
Please give some comments on each arm.

Reply:
this is for reading bits according to pci spec. I have added a summarised comment above
"Support reading capability id, next capability pointer, mask bits and pending bit."